provider/digitalocean/digital_ocean.go

```diff
@@ -179,160 +236,379 @@ func (p *DigitalOceanProvider) fetchZones(ctx context.Context) ([]godo.Domain, e
+					RecordID: record.ID,
+				})
+			}
 		}
 	}
+
 	return nil
 }
 
-// ApplyChanges applies a given set of changes in a given zone.
-func (p *DigitalOceanProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
-	combinedChanges := make([]*DigitalOceanChange, 0, len(changes.Create)+len(changes.UpdateNew)+len(changes.Delete))
+func processDeleteActions(
+	recordsByDomain map[string][]godo.DomainRecord,
+	deletesByDomain map[string][]*endpoint.Endpoint,
+	changes *digitalOceanChanges,
+) error {
+	// Generate delete actions for each deleted endpoint.
+	for domain, deletes := range deletesByDomain {
+		if len(deletes) == 0 {
+			log.WithFields(log.Fields{
+				"domain": domain,
+			}).Debug("Skipping Zone, no deletes found.")
+			continue
+		}
 
-	combinedChanges = append(combinedChanges, newDigitalOceanChanges(DigitalOceanCreate, changes.Create)...)
-	combinedChanges = append(combinedChanges, newDigitalOceanChanges(DigitalOceanUpdate, changes.UpdateNew)...)
-	combinedChanges = append(combinedChanges, newDigitalOceanChanges(DigitalOceanDelete, changes.Delete)...)
+		records := recordsByDomain[domain]
 
-	return p.submitChanges(ctx, combinedChanges)
-}
+		for _, ep := range deletes {
+			matchingRecords := getMatchingDomainRecords(records, domain, ep)
+
+			if len(matchingRecords) == 0 {
+				log.WithFields(log.Fields{
+					"domain":     domain,
+					"dnsName":    ep.DNSName,
+					"recordType": ep.RecordType,
+				}).Warn("Records to delete not found.")
+			}
 
-// newDigitalOceanChanges returns a collection of Changes based on the given records and action.
-func newDigitalOceanChanges(action string, endpoints []*endpoint.Endpoint) []*DigitalOceanChange {
-	changes := make([]*DigitalOceanChange, 0, len(endpoints))
+			for _, record := range matchingRecords {
+				doDelete := false
+				for _, t := range ep.Targets {
+					v1 := t
+					v2 := record.Data
+					if ep.RecordType == endpoint.RecordTypeCNAME {
+						v1 = strings.TrimSuffix(t, ".")
+						v2 = strings.TrimSuffix(t, ".")
+					}
+					if v1 == v2 {
+						doDelete = true
+					}
+				}
 
-	for _, endpoint := range endpoints {
-		changes = append(changes, newDigitalOceanChange(action, endpoint))
+				if doDelete {
+					changes.Deletes = append(changes.Deletes, &digitalOceanChangeDelete{
+						Domain:   domain,
+						RecordID: record.ID,
+					})
+				}
+			}
+		}
 	}
 
-	return changes
+	return nil
 }
 
-func newDigitalOceanChange(action string, endpoint *endpoint.Endpoint) *DigitalOceanChange {
-	// no annotation results in a TTL of 0, default to 300 for consistency with other providers
-	var ttl = digitalOceanRecordTTL
-	if endpoint.RecordTTL.IsConfigured() {
-		ttl = int(endpoint.RecordTTL)
+// ApplyChanges applies the given set of generic changes to the provider.
+func (p *DigitalOceanProvider) ApplyChanges(ctx context.Context, planChanges *plan.Changes) error {
+	// TODO: This should only retrieve zones affected by the given `planChanges`.
```

Comment:
Should we have this done before merging this PR or do you want to keep it like this? What are the consequences?

Reply:
Not implementing this TODO would leave the status quo in place which is that we would just retrieve all of the records that `Records` retrieves, just as the existing version of the provider does. (`--domain-filter` would still be applied to limit the provider to only those zones where external-dns is managing records.)

My concern here was for avoiding having to retrieve all of the records if the zone had potentially many records. https://github.com/kubernetes-sigs/external-dns/issues/1429 was on my mind as it flagged too many API calls as a potential hazard in such cases.

This TODO is a nice to have but should not block landing this PR.